ios/uPortMobile/Classes/EthereumSigner.m

```diff
@@ -0,0 +1,246 @@
+  if (!BN_mul_word(x, i)) { ret=-1; goto err; }
+  if (!BN_add(x, x, r)) { ret=-1; goto err; }
+  field = BN_CTX_get(ctx);
+  if (!EC_GROUP_get_curve_GFp(group, field, NULL, NULL, ctx)) { ret=-2; goto err; }
+  if (BN_cmp(x, field) >= 0) { ret=0; goto err; }
+  if ((R = EC_POINT_new(group)) == NULL) { ret = -2; goto err; }
+  if (!EC_POINT_set_compressed_coordinates_GFp(group, R, x, recid % 2, ctx)) { ret=0; goto err; }
+  if (check) {
+    if ((O = EC_POINT_new(group)) == NULL) { ret = -2; goto err; }
+    if (!EC_POINT_mul(group, O, NULL, R, order, ctx)) { ret=-2; goto err; }
+    if (!EC_POINT_is_at_infinity(group, O)) { ret = 0; goto err; }
+  }
+  if ((Q = EC_POINT_new(group)) == NULL) { ret = -2; goto err; }
+  n = EC_GROUP_get_degree(group);
+  e = BN_CTX_get(ctx);
+  if (!BN_bin2bn(msg, msglen, e)) { ret=-1; goto err; }
+  if (8*msglen > n) BN_rshift(e, e, 8-(n & 7));
+  zero = BN_CTX_get(ctx);
+  if (!BN_zero(zero)) { ret=-1; goto err; }
+  if (!BN_mod_sub(e, zero, e, order, ctx)) { ret=-1; goto err; }
+  rr = BN_CTX_get(ctx);
+  if (!BN_mod_inverse(rr, r, order, ctx)) { ret=-1; goto err; }
+  sor = BN_CTX_get(ctx);
+  if (!BN_mod_mul(sor, s, rr, order, ctx)) { ret=-1; goto err; }
+  eor = BN_CTX_get(ctx);
+  if (!BN_mod_mul(eor, e, rr, order, ctx)) { ret=-1; goto err; }
+  if (!EC_POINT_mul(group, Q, eor, R, sor, ctx)) { ret=-2; goto err; }
+  if (!EC_KEY_set_public_key(eckey, Q)) { ret=-2; goto err; }
+  
+  ret = 1;
+  
+err:
+  if (ctx) {
+    BN_CTX_end(ctx);
+    BN_CTX_free(ctx);
+  }
+  if (R != NULL) EC_POINT_free(R);
+  if (O != NULL) EC_POINT_free(O);
+  if (Q != NULL) EC_POINT_free(Q);
+  return ret;
+}
+
+NSMutableData *compressedPublicKey(EC_KEY *key) {
+  if (!key) return nil;
+  EC_KEY_set_conv_form(key, POINT_CONVERSION_COMPRESSED);
+  int length = i2o_ECPublicKey(key, NULL);
+  if (!length) return nil;
+  NSCAssert(length <= 65, @"Pubkey length must be up to 65 bytes.");
+  NSMutableData* data = [[NSMutableData alloc] initWithLength:length];
+  unsigned char* bytes = [data mutableBytes];
+  if (i2o_ECPublicKey(key, &bytes) != length) return nil;
+  return data;
+}
+
+
+NSDictionary *ethereumSignature(BTCKey *keypair, NSData *hash, NSData *chainId) {
+  NSDictionary *sig = genericSignature(keypair, hash, YES);
+  NSData *rData = (NSData *)sig[@"r"];
+  NSData *sData = (NSData *)sig[@"s"];
+  BN_ULONG base = 0x1b; // pre-EIP155
+  if (chainId) {
+    BIGNUM *v = BN_new(); BN_bin2bn(chainId.bytes, chainId.length, v);
+    // TODO support longer chainIDs
+    base = BN_get_word(v) * 2 + 35;
+    BN_clear_free(v);
+  }
+  
+  NSDictionary *signatureDictionary = @{ @"v": @(base + [sig[@"recoveryParam"] intValue]),
+                                         @"r": [rData base64EncodedStringWithOptions:0],
+                                         @"s":[sData base64EncodedStringWithOptions:0]};
+  return signatureDictionary;
+}
+
+
+NSDictionary *genericSignature(BTCKey *keypair, NSData *hash, BOOL lowS) {
```

Comment:
@christianlundkvist @mirceanis This code occasionally creates an invalid signature. It's something like 1 in a 1000 (unscientifically).

Can you guys please look at the maths if there is something wrong I'm doing.

Reply:
@pelle do you have an example of invalid signature?